sourcegraph/sourcegraph
# Code Insights: Fix voice over for creation ui complex field validation checks

Closes https://github.com/sourcegraph/sourcegraph/issues/41672

## Background
In this PR we fix problems with validation over search box query input. Prior to this PR non of validation check that are placed below the input didn't work (hadn't been announced by screen reader), this PR fixes and also changes the way how we provide data for those checks. Previously, we used simple selectors inside of valid function to mark input as invalid AND in render call tree in order to mark checks UI with valid/invalid (green and red marks) UI. 

This logic wasn't optimal because parsing query isn't necessary an easy work to do, and this forced us to track calling searchQueryChecks in more than just one place. 

In this PR we have only one place to carry about validation and it's validator function all other UI is dependent on validation state. 

**Small fix**: This PR also fixes small problem with search box code mirror layout (bad line-height value)

https://user-images.githubusercontent.com/18492575/194138327-71187654-4808-4a35-aa7c-83b6255962a1.mov

## Problems that we still have 
- Sometimes (for some reason and I'm still looking why) it's possible to hear failed validation check twice 
- We don't have control over order of role=alert announcements and at the moment in progress, because of this "Query is not valid" is announced after specific failed validation check. It would be better if we put "Query is not valid" first.

## Test plan
- Make sure that logic over capture group validation check works as expected
- Make sure that if you have value in the search box field and you refresh the page you can see validation state immediately 



## App preview:

- [Web](https://sg-web-vk-fix-capture-validation-check.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-dfsidrptlj.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.
